File: heapsort/heapsort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Функция для восстановления свойств кучи
void heapify(vector<int>& arr, int n, int i) {
    int largest = i;       // Инициализируем корень как наибольший элемент
    int left = 2 * i + 1;  // Левый дочерний элемент
    int right = 2 * i + 2; // Правый дочерний элемент

    // Если левый дочерний элемент больше корня
    if (left < n && arr[left] > arr[largest])
        largest = left;

    // Если правый дочерний элемент больше текущего наибольшего
    if (right < n && arr[right] > arr[largest])
        largest = right;

    // Если наибольший элемент не корень, выполняем обмен
    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest); // Рекурсивно восстанавливаем кучу
    }
}

// Основная функция сортировки Heapsort
void heapsort(vector<int>& arr) {
    int n = arr.size();

    // Построение кучи (перестановка массива)
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);

    // Извлечение элементов из кучи
    for (int i = n - 1; i >= 0; i--) {
        swap(arr[0], arr[i]);     // Перемещаем текущий корень в конец
        heapify(arr, i, 0);       // Восстанавливаем свойства кучи для оставшихся элементов
    }
}

#ifndef TESTING  // Условная компиляция: main() будет включен, только если TESTING не определён
int main() {
    int n;
    cout << "Введите количество элементов: ";
    cin >> n;

    vector<int> arr(n);
    cout << "Введите элементы массива:\n";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    heapsort(arr);

    cout << "Отсортированный массив:\n";
    for (int x : arr) {
        cout << x << " ";
    }
    cout << endl;

    return 0;
}
#endif

File: heapsort/heapsort.py
def heapify(arr, n, i):
    largest = i       # Инициализируем наибольший элемент как корень
    left = 2 * i + 1  # левый = 2*i + 1
    right = 2 * i + 2 # правый = 2*i + 2

    # Проверяем, существует ли левый дочерний элемент и больше ли он корня
    if left < n and arr[left] > arr[largest]:
        largest = left

    # Проверяем, существует ли правый дочерний элемент и больше ли он текущего наибольшего
    if right < n and arr[right] > arr[largest]:
        largest = right

    # Если наибольший элемент не корень
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # обмен

        # Рекурсивно применяем heapify к поддереву
        heapify(arr, n, largest)

def heapsort(arr):
    n = len(arr)

    # Построение максимальной кучи
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Один за другим извлекаем элементы из кучи
    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]   # обмен
        heapify(arr, i, 0)

# Пример использования
if __name__ == "__main__":
    data = [12, 11, 13, 5, 6, 7]
    print("Исходный массив:", data)
    heapsort(data)
    print("Отсортированный массив:", data)